Fix missing translations in inspector

tarhi-saad opened the pull request:
This PR fixes the missing translations in the inspector (in the Editor mode) of the C&C Blocks & Inner Blocks!

Unlike standard blocks, the translations for the `Inner Blocks` (e.g., `title`, `description` & `keywords`) doesn't work when the [blocks are registered on the client-side only](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-registration.md). We are still unsure what is causing this unexpected behavior. Even though Gutenberg is [handling i18n for client-side only registered blocks](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#javascript), we didn't find a way to make it work in our case.

The expected behavior was for Gutenberg to wrap all localized properties in `_x` (from `@wordpress/i18n package`) function calls, similar to how it works in PHP. And then for [GlotPress](https://glotpress.blog/) to generate `.json` files in the `wp-content/languages/` folder containing the corresponding translations.

On the other hand, our missing translations can be found inside the `.po` files in the `wp-content/languages/` folder. These files are generated by [GlotPress](https://glotpress.blog/) when translating server-side (PHP). After this realization, the solution became clear: register our `Inner Blocks` server-side instead! This was also an early suggestion made by @senadir! Besides, the [recommended method to register new block types](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md) is to use the `block.json` metadata file (i.e., register the block server-side). So, it was time for the implementation 💪:

And indeed, after adopting this approach for the `Inner Blocks`, the missing translations bug was fixed. The [translation was handled by WordPress](https://github.com/WordPress/WordPress/blob/35ab93306779cf1b04eba1b614f382365bea0b43/wp-includes/blocks.php#L308) on the back-end. [WordPress string discovery system can automatically translate](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#internationalization) fields marked in the `block.json` file as translatable.

Some changes that needed to be done to properly register our `Inner Blocks` server-side:

- [Generate block.json files for Inner Blocks](https://github.com/woocommerce/woocommerce-blocks/blob/b220e4937665119c139a01f43f0149111c95444b/bin/webpack-configs.js#L264-L277)
- [Register only the client-side settings on the client](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#javascript-client-side).
- Prevent WordPress from [injecting script tags](https://github.com/WordPress/WordPress/blob/master/wp-includes/blocks.php#L314-L316) for each Inner Block registered server-side by omitting the [editor_script](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#editor-script) field from our [register_block_type](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTypes/AbstractBlock.php#L201) function and creating the new [AbstractInnerBlock](https://github.com/woocommerce/woocommerce-blocks/blob/b220e4937665119c139a01f43f0149111c95444b/src/BlockTypes/AbstractInnerBlock.php) for this purpose.

**Things that need improvement:**

- [x] Currently, the `/build` folder is [polluted with so many top-level folders](https://github.com/woocommerce/woocommerce-blocks/pull/6737#discussion_r933517699). To fix this, we can choose to generate the `Inner Blocks` `blocks.json` files inside a parent folder (e.g., `/checkout-blocks` and `/cart-blocks`).

 
### TODO

- [x] Fix the translation for the Checkout's Inner Blocks

<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

Fixes #6244

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) .
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md)
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Change your site language (I tested using French & NL)
2. Go to the C&C Blocks (Editor mode)
3. Select each inner block and make sure the `title` & `description` are translated in the inspector. See image below:

![image](https://user-images.githubusercontent.com/14235870/181719689-f9decbdd-5bc6-4192-8612-4b2696a8d7c4.png)

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Fix missing translations in the inspector (editor mode)

tarhi-saad committed:
Try registering the "cart taxes" inner block

Registering server side. This example isn't working, but I'm pushing to
share it and see what's wrong with this implementation.

tarhi-saad committed:
Fix registering the cart taxes inner block issue

tarhi-saad committed:
Update translation script loading

senadir commented on assets/js/blocks/cart/inner-blocks/cart-order-summary-taxes/index.tsx:
```diff
@@ -21,7 +24,20 @@ registerFeaturePluginBlockType( metadata, {
+	// 	'Shows the cart taxes row.',
+	// 	'block description',
+	// 	'woo-gutenberg-products-block'
+	// ),
```
All of those information should be present in the metadata file, why are you rewriting them?

tarhi-saad replied:
I'm overwriting the `title` & `description` to test wrapping them with the `@wordpress/i18n` functions and see if it would make any change. This worked for the `title` (it was translated correctly), but it didn't work for the `description` since the corresponding translation script/file wasn't loaded.

tarhi-saad replied:
Seeing other blocks using the `block.json` file (e.g. `products-by-attribute`), I think we shouldn't need the `@wordpress/i18n` functions for our translation to work! 🙌

senadir commented on src/BlockTypesController.php:
```diff
@@ -201,6 +201,7 @@ protected function get_block_types() {
 		if ( Package::feature()->is_feature_plugin_build() ) {
 			$block_types[] = 'Checkout';
 			$block_types[] = 'Cart';
+			$block_types[] = 'CartOrderSummaryTaxesBlock';
```
We might want to have this logic inside Cart.php class?

tarhi-saad replied:
✅ Fixed in: [83a8b43](https://github.com/woocommerce/woocommerce-blocks/pull/6737/commits/83a8b433fe749017f330920db89ab70501426037)

senadir reviewed (commented):
I left a couple of questions while I dig more into this.

tarhi-saad committed:
Remove unnecessary JS translation

The translation should work fine by getting the title & description from
the `block.json` file

senadir commented on bin/webpack-entries.js:
```diff
@@ -45,6 +45,9 @@ const blocks = {
 	'stock-filter': {},
 	'active-filters': {},
 	cart: {},
+	'cart-order-summary-taxes-block': {
+		customDir: 'cart/inner-blocks/cart-order-summary-taxes',
+	},
```
I question the need for this one and maybe we can just rely on cart.js file :/

tarhi-saad replied:
When I load the page (the cart edit mode), I get this error:

<img width="616" alt="image" src="https://user-images.githubusercontent.com/14235870/181003177-bb937323-a49a-4ee1-bed5-8c2fdcc2914c.png">

And after adding the code above, it stopped showing this console error. I think now that the `taxes'` inner block is registered server side, it's expecting to have its own `JS` file. But I'll look into that and see If we can choose not to include a JS file in this case.

tarhi-saad replied:
✅ Fixed here: [c5670f6](https://github.com/woocommerce/woocommerce-blocks/pull/6737/commits/c5670f6be17c2de2552028dc959bbcb8f4e57055)

tarhi-saad committed:
Put back the initial code in the 'Cart Taxes' inner block

We didn't provide the correct `block.json` file path server side,
that's why the `metadata` wasn't correctly registered

tarhi-saad committed:
Generate `block.json`files for inner blocks

This is the first step on fixing the missing translations of `metadata`
in `block.json` files

tarhi-saad committed:
Set the folder name exactly the same as the inner block name

We are doing this first test for the `Cart taxes` inner block.

The `Block` & its containing folder need to have the same name for:
- Consistency
- We use the `Block` name to get the file Path

tarhi-saad committed:
Update imports after folder renaming

tarhi-saad committed:
Get block name directly from the JSON metadata

Getting the block name from the JSON metadata is less error prone than
extracting it from the file path.

And no need to rename all our `inner-blocks` to get the correct
`block.json` path

tarhi-saad committed:
Revert folder naming change of `Cart taxes` inner block

Since we are getting the `block` name directly from the `block.json`
metadata instead of extracting it from the file path, there is no need
to keep their names in sync anymore

tarhi-saad committed:
Fix missing translations for the `Cart Subtotal` Block

tarhi-saad committed:
Register only the client-side settings on the client

When the block is registered on the server, you only need to register
the client-side settings on the client using the same block’s name.
See [docs](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#javascript-client-side).

tarhi-saad committed:
Add schema validation to `block.json`

Development is improved by using a defined schema definition file.
Supported editors can provide help like tooltips, autocomplete, and
schema validation.

tarhi-saad committed:
Use the same `editor_script` as the parent block

This prevents WordPress from generating script tags to inexistant
inner blocks JS files

tarhi-saad committed:
Add C&C inner blocks in Cart.php & Checkout.php

This is a refactoring to keep the block types controller file less
overloaded

tarhi-saad committed:
Fix all Cart inner blocks missing translations

senadir commented on bin/webpack-configs.js:
```diff
@@ -262,14 +263,19 @@ const getMainConfig = ( options = {} ) => {
 						from: './assets/js/blocks/**/block.json',
 						to( { absoluteFilename } ) {
-							const blockName = absoluteFilename
+							/**
+							 * Getting the block name from the JSON metadata is less error prone
+							 * than extracting it from the file path.
+							 */
+							const JSONFile = fs.readFileSync(
+								path.resolve( __dirname, absoluteFilename )
+							);
+							const metadata = JSON.parse( JSONFile.toString() );
+							const blockName = metadata.name
 								.split( '/' )
-								.at( -2 );
+								.at( 1 );
 							return `./${ blockName }/block.json`;
 						},
```
I assume we have little control into where this goes? currently, it pollutes our `build` folder with so many top level folders.
<img width="352" alt="image" src="https://user-images.githubusercontent.com/6165348/181820832-0e1284a0-a250-417d-818e-83f420cd9de2.png">


We actually have two folders there `checkout-blocks` and `cart-blocks`. Do you think we can output those block.json files there, and how much effort would it require? For block.json, we can change that in the future so this is not a blocker.

senadir reviewed (approved):
I tested this and it's working fine! Thank you Saad for spending time with this.

tarhi-saad committed:
Create the "AbstractInnerBlock" class

The "Inner Blocks" will use their parent's script, so no need to create
new scripts for each one of them

And, our "Inner Blocks" should always be registered using the metadata file

tarhi-saad committed:
Update the "Inner Blocks" PHP classes

tarhi-saad committed:
Fix PHP lint erros & update function description

tarhi-saad committed:
Fix missing translations bug for all Checkout Inner Blocks

tarhi-saad committed:
Merge branch 'trunk' into fix/6244-missing-translations-in-inspector

senadir commented on assets/js/blocks/cart/inner-blocks/cart-accepted-payment-methods-block/block.json:
```diff
@@ -13,5 +13,6 @@
 	},
 	"parent": [ "woocommerce/cart-totals-block" ],
 	"textdomain": "woo-gutenberg-products-block",
+	"$schema": "https://schemas.wp.org/trunk/block.json",
```
I forget to ask why was this needed?

tarhi-saad replied:
Loading the [schema definition ](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md#benefits-using-the-metadata-file) file allows supported editors to provide help like tooltips, autocomplete, and schema validation. For example, you can hover over a field in `block.json` to see a tooltip description for it. So, it's always good to add this line for development purposes.

senadir commented on src/BlockTypes/Checkout.php:
```diff
@@ -411,4 +413,39 @@ protected function register_block_type_assets() {
+		$block_types = [];
+
+		if ( Package::feature()->is_feature_plugin_build() ) {
+			$block_types[] = 'Checkout';
+			$block_types[] = 'Checkout';
```
Checkout is twice here
```suggestion
			$block_types[] = 'Checkout';
```

senadir replied:
This is throwing an error for me.

tarhi-saad replied:
Thank you, @senadir! 💯

senadir reviewed (approved):
This is looking good, I left a single comment but it's ready to merge.

tarhi-saad committed:
Update src/BlockTypes/Checkout.php

Co-authored-by: Seghir Nadir <nadir.seghir@gmail.com>

senadir reviewed (changes requested):
I was actually too quick to approve this without testing, I'm still seeing issues on frontend with scripts like `/proceed-to-checkout-block-frontend.js` missing.

senadir committed:
skip lazy loaded scripts

senadir reviewed (approved):
This is now working fine without issues.
